### ISSUE TITLE
Confusing error message for multiply defined variables

### ISSUE THREAD

**Issue by jsmall-nvidia** [shader-slang/slang#1077]
Take the following shader, where 'stride' is multiply defined in error...

```
RWStructuredBuffer<float3> positions;

[numthreads(256, 1, 1)]
void computeMain(uint3 globalIdx : SV_DispatchThreadID)
{              
    uint count, stride;
    positions.GetDimensions(count, stride);
 
    int stride = 10;
}
```
On slang the error reported is 

```
D:\local\Temp\a335d33f-98c1-452f-bc4d-176733908e9c.hlsl(7): error 30019: expected an expression of type 'uint', got 'overload group'
D:\local\Temp\a335d33f-98c1-452f-bc4d-176733908e9c.hlsl(7): error 30047: argument passed to parameter '1' must be l-value.
D:\local\Temp\a335d33f-98c1-452f-bc4d-176733908e9c.hlsl(7): note 30048: argument was implicitly cast from 'overload group' to 'error', and Slang does not support using an implicit cast as an l-value
```

That's quite a confusing message - far more so than 'multiply defined variable' or something of that kind. Also it does not list the line ` int stride = 10;` which is pretty important in diagnosing the issue. If there were 100s of lines before that line - the message focuses on only the earlier usage, which would be hard to figure out whats going on. 

By contrast dxc outputs 

```
D:\local\Temp\b2dcd9d2-65c9-4683-8103-f281751c9625.hlsl:9:9: error: redefinition of 'stride' with a different type: 'int' vs 'uint'
    int stride = 10;
        ^
D:\local\Temp\b2dcd9d2-65c9-4683-8103-f281751c9625.hlsl:6:17: note: previous definition is here
    uint count, stride;
```

Which highlights the lines of significance and has a much clearer error message.

**Comment by tfoleyNV** [comment]
Yes, this is a pretty bad error message on our part and is a quality-of-life thing we should find time to fix.

The basic problem is that `slang-check.cpp` currently doesn't proactively check for duplicate/conflicting definitions (at least not to the extent it should), and instead relies on just emitting an error in cases where those declarations are used ambiguously. In this case, the ambiguous use gets wrapped up in another error message about what we pass to `out`/`intout` parameters. It is also tied up in the way that our approach that tries to parse and *then* apply semantic checking means that the call site for `GetDimensions()` there can "see" the later declaration of `stride`, even if it doesn't really make sense.

**Comment by tfoleyNV** [comment]
This should be fixed with #1201.
